File: combinatorics/resource/LCHP/problem_LTDT_code/bai_7/bai7.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <algorithm>

using namespace std;

// Cấu trúc node cho cây nhị phân
struct TreeNode {
    int data;
    TreeNode* left;
    TreeNode* right;
    
    TreeNode(int val) : data(val), left(nullptr), right(nullptr) {}
};

// Cấu trúc node cho cây tổng quát
struct GeneralTreeNode {
    int data;
    vector<GeneralTreeNode*> children;
    
    GeneralTreeNode(int val) : data(val) {}
};

class TreeTraversal {
public:
    // ===== PREORDER TRAVERSAL =====
    
    // Đệ quy
    void preorderRecursive(TreeNode* root, vector<int>& result) {
        if (root == nullptr) return;
        
        result.push_back(root->data);           // Thăm gốc
        preorderRecursive(root->left, result);  // Duyệt trái
        preorderRecursive(root->right, result); // Duyệt phải
    }
    
    // Không đệ quy
    vector<int> preorderIterative(TreeNode* root) {
        vector<int> result;
        if (root == nullptr) return result;
        
        stack<TreeNode*> stk;
        stk.push(root);
        
        while (!stk.empty()) {
            TreeNode* node = stk.top();
            stk.pop();
            
            result.push_back(node->data);
            
            // Push phải trước, trái sau (vì stack là LIFO)
            if (node->right) stk.push(node->right);
            if (node->left) stk.push(node->left);
        }
        
        return result;
    }
    
    // ===== POSTORDER TRAVERSAL =====
    
    // Đệ quy
    void postorderRecursive(TreeNode* root, vector<int>& result) {
        if (root == nullptr) return;
        
        postorderRecursive(root->left, result);  // Duyệt trái
        postorderRecursive(root->right, result); // Duyệt phải
        result.push_back(root->data);            // Thăm gốc
    }
    
    // Không đệ quy - Phương pháp 2 stack
    vector<int> postorderIterative(TreeNode* root) {
        vector<int> result;
        if (root == nullptr) return result;
        
        stack<TreeNode*> stk1, stk2;
        stk1.push(root);
        
        // Stack 1: duyệt và đẩy vào stack 2
        while (!stk1.empty()) {
            TreeNode* node = stk1.top();
            stk1.pop();
            stk2.push(node);
            
            if (node->left) stk1.push(node->left);
            if (node->right) stk1.push(node->right);
        }
        
        // Stack 2: lấy ra theo thứ tự postorder
        while (!stk2.empty()) {
            result.push_back(stk2.top()->data);
            stk2.pop();
        }
        
        return result;
    }
    
    // Không đệ quy - Phương pháp 1 stack với flag
    vector<int> postorderIterativeSingleStack(TreeNode* root) {
        vector<int> result;
        if (root == nullptr) return result;
        
        stack<pair<TreeNode*, bool>> stk; // bool: đã xử lý con hay chưa
        stk.push({root, false});
        
        while (!stk.empty()) {
            auto [node, processed] = stk.top();
            stk.pop();
            
            if (processed) {
                result.push_back(node->data);
            } else {
                // Đánh dấu đã xử lý và push lại
                stk.push({node, true});
                // Push con phải trước, trái sau
                if (node->right) stk.push({node->right, false});
                if (node->left) stk.push({node->left, false});
            }
        }
        
        return result;
    }
    
    // ===== TOP-DOWN TRAVERSAL =====
    
    vector<vector<int>> topDownTraversal(TreeNode* root) {
        vector<vector<int>> result;
        if (root == nullptr) return result;
        
        queue<TreeNode*> q;
        q.push(root);
        
        while (!q.empty()) {
            int levelSize = q.size();
            vector<int> currentLevel;
            
            for (int i = 0; i < levelSize; i++) {
                TreeNode* node = q.front();
                q.pop();
                
                currentLevel.push_back(node->data);
                
                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
            
            result.push_back(currentLevel);
        }
        
        return result;
    }
    
    // Phiên bản chỉ trả về 1 vector
    vector<int> topDownFlat(TreeNode* root) {
        vector<int> result;
        if (root == nullptr) return result;
        
        queue<TreeNode*> q;
        q.push(root);
        
        while (!q.empty()) {
            TreeNode* node = q.front();
            q.pop();
            
            result.push_back(node->data);
            
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }
        
        return result;
    }
    
    // ===== BOTTOM-UP TRAVERSAL =====
    
    vector<vector<int>> bottomUpTraversal(TreeNode* root) {
        vector<vector<int>> levels = topDownTraversal(root);
        reverse(levels.begin(), levels.end());
        return levels;
    }
    
    vector<int> bottomUpFlat(TreeNode* root) {
        vector<int> result;
        
        // Chuyển đổi thành bottom-up bằng cách group theo level rồi reverse
        vector<vector<int>> levels = topDownTraversal(root);
        
        for (int i = levels.size() - 1; i >= 0; i--) {
            for (int val : levels[i]) {
                result.push_back(val);
            }
        }
        
        return result;
    }
    
    // ===== CÂY TỔNG QUÁT =====
    
    // Preorder cho cây tổng quát
    void preorderGeneral(GeneralTreeNode* root, vector<int>& result) {
        if (root == nullptr) return;
        
        result.push_back(root->data);
        for (GeneralTreeNode* child : root->children) {
            preorderGeneral(child, result);
        }
    }
    
    // Postorder cho cây tổng quát
    void postorderGeneral(GeneralTreeNode* root, vector<int>& result) {
        if (root == nullptr) return;
        
        for (GeneralTreeNode* child : root->children) {
            postorderGeneral(child, result);
        }
        result.push_back(root->data);
    }
    
    // Level order cho cây tổng quát
    vector<vector<int>> levelOrderGeneral(GeneralTreeNode* root) {
        vector<vector<int>> result;
        if (root == nullptr) return result;
        
        queue<GeneralTreeNode*> q;
        q.push(root);
        
        while (!q.empty()) {
            int levelSize = q.size();
            vector<int> currentLevel;
            
            for (int i = 0; i < levelSize; i++) {
                GeneralTreeNode* node = q.front();
                q.pop();
                
                currentLevel.push_back(node->data);
                
                for (GeneralTreeNode* child : node->children) {
                    q.push(child);
                }
            }
            
            result.push_back(currentLevel);
        }
        
        return result;
    }
    
    // ===== UTILITY FUNCTIONS =====
    
    // Tạo cây mẫu
    TreeNode* createSampleTree() {
        /*
                1
               / \
              2   3
             / \   \
            4   5   6
        */
        TreeNode* root = new TreeNode(1);
        root->left = new TreeNode(2);
        root->right = new TreeNode(3);
        root->left->left = new TreeNode(4);
        root->left->right = new TreeNode(5);
        root->right->right = new TreeNode(6);
        return root;
    }
    
    GeneralTreeNode* createSampleGeneralTree() {
        /*
                1
              / | \
             2  3  4
            /|    |\
           5 6    7 8
        */
        GeneralTreeNode* root = new GeneralTreeNode(1);
        root->children.push_back(new GeneralTreeNode(2));
        root->children.push_back(new GeneralTreeNode(3));
        root->children.push_back(new GeneralTreeNode(4));
        
        root->children[0]->children.push_back(new GeneralTreeNode(5));
        root->children[0]->children.push_back(new GeneralTreeNode(6));
        
        root->children[2]->children.push_back(new GeneralTreeNode(7));
        root->children[2]->children.push_back(new GeneralTreeNode(8));
        
        return root;
    }
    
    // In kết quả
    void printVector(const vector<int>& vec, const string& name) {
        cout << name << ": ";
        for (int val : vec) {
            cout << val << " ";
        }
        cout << endl;
    }
    
    void printLevels(const vector<vector<int>>& levels, const string& name) {
        cout << name << ":" << endl;
        for (int i = 0; i < levels.size(); i++) {
            cout << "Level " << i << ": ";
            for (int val : levels[i]) {
                cout << val << " ";
            }
            cout << endl;
        }
    }
    
    // Tính chiều cao cây
    int getHeight(TreeNode* root) {
        if (root == nullptr) return 0;
        return 1 + max(getHeight(root->left), getHeight(root->right));
    }
    
    // Đếm số node
    int countNodes(TreeNode* root) {
        if (root == nullptr) return 0;
        return 1 + countNodes(root->left) + countNodes(root->right);
    }
    
    // Kiểm tra cây có cân bằng hay không
    bool isBalanced(TreeNode* root) {
        return checkBalance(root) != -1;
    }
    
private:
    int checkBalance(TreeNode* root) {
        if (root == nullptr) return 0;
        
        int leftHeight = checkBalance(root->left);
        if (leftHeight == -1) return -1;
        
        int rightHeight = checkBalance(root->right);
        if (rightHeight == -1) return -1;
        
        if (abs(leftHeight - rightHeight) > 1) return -1;
        
        return 1 + max(leftHeight, rightHeight);
    }
};

// Hàm test toàn bộ
int main() {
    TreeTraversal traverser;
    
    cout << "=== DEMO DUYỆT CÂY NHỊ PHÂN ===" << endl;
    
    // Tạo cây mẫu
    TreeNode* root = traverser.createSampleTree();
    
    cout << "Cây mẫu:" << endl;
    cout << "        1" << endl;
    cout << "       / \\" << endl;
    cout << "      2   3" << endl;
    cout << "     / \\   \\" << endl;
    cout << "    4   5   6" << endl;
    cout << endl;
    
    // Test preorder
    vector<int> preResult;
    traverser.preorderRecursive(root, preResult);
    traverser.printVector(preResult, "Preorder (Đệ quy)");
    
    vector<int> preIter = traverser.preorderIterative(root);
    traverser.printVector(preIter, "Preorder (Không đệ quy)");
    
    // Test postorder
    vector<int> postResult;
    traverser.postorderRecursive(root, postResult);
    traverser.printVector(postResult, "Postorder (Đệ quy)");
    
    vector<int> postIter = traverser.postorderIterative(root);
    traverser.printVector(postIter, "Postorder (2 stack)");
    
    vector<int> postSingle = traverser.postorderIterativeSingleStack(root);
    traverser.printVector(postSingle, "Postorder (1 stack)");
    
    // Test level order
    vector<vector<int>> topDown = traverser.topDownTraversal(root);
    traverser.printLevels(topDown, "Top-down traversal");
    
    vector<vector<int>> bottomUp = traverser.bottomUpTraversal(root);
    traverser.printLevels(bottomUp, "Bottom-up traversal");
    
    cout << endl;
    
    // Thông tin cây
    cout << "=== THÔNG TIN CÂY ===" << endl;
    cout << "Chiều cao: " << traverser.getHeight(root) << endl;
    cout << "Số node: " << traverser.countNodes(root) << endl;
    cout << "Cân bằng: " << (traverser.isBalanced(root) ? "Có" : "Không") << endl;
    
    cout << endl;
    
    // Test cây tổng quát
    cout << "=== DEMO DUYỆT CÂY TỔNG QUÁT ===" << endl;
    
    GeneralTreeNode* genRoot = traverser.createSampleGeneralTree();
    
    cout << "Cây tổng quát mẫu:" << endl;
    cout << "        1" << endl;
    cout << "      / | \\" << endl;
    cout << "     2  3  4" << endl;
    cout << "    /|    |\\" << endl;
    cout << "   5 6    7 8" << endl;
    cout << endl;
    
    vector<int> genPre, genPost;
    traverser.preorderGeneral(genRoot, genPre);
    traverser.postorderGeneral(genRoot, genPost);
    
    traverser.printVector(genPre, "Preorder cây tổng quát");
    traverser.printVector(genPost, "Postorder cây tổng quát");
    
    vector<vector<int>> genLevels = traverser.levelOrderGeneral(genRoot);
    traverser.printLevels(genLevels, "Level order cây tổng quát");
    
    return 0;
}

File: combinatorics/resource/LCHP/problem_LTDT_code/bai_7/bai7.py
from collections import deque
from typing import List, Optional

class TreeNode:
    """Node cho cây nhị phân"""
    def __init__(self, val: int = 0):
        self.val = val
        self.left: Optional['TreeNode'] = None
        self.right: Optional['TreeNode'] = None

class GeneralTreeNode:
    """Node cho cây tổng quát"""
    def __init__(self, val: int = 0):
        self.val = val
        self.children: List['GeneralTreeNode'] = []

class TreeTraversal:
    """Lớp xử lý các phương pháp duyệt cây"""
    
    # ===== PREORDER TRAVERSAL =====
    
    def preorder_recursive(self, root: Optional[TreeNode]) -> List[int]:
        """
        Duyệt preorder bằng đệ quy
        Thứ tự: Gốc -> Trái -> Phải
        """
        result = []
        
        def dfs(node):
            if not node:
                return
            result.append(node.val)      # Thăm gốc
            dfs(node.left)            # Duyệt trái
            dfs(node.right)           # Duyệt phải
        
        dfs(root)
        return result
    
    def preorder_iterative(self, root: Optional[TreeNode]) -> List[int]:
        """
        Duyệt preorder không đệ quy bằng stack
        """
        if not root:
            return []
        
        result = []
        stack = [root]
        
        while stack:
            node = stack.pop()
            result.append(node.val)
            
            # Push phải trước, trái sau (vì stack là LIFO)
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
        
        return result
    
    # ===== POSTORDER TRAVERSAL =====
    
    def postorder_recursive(self, root: Optional[TreeNode]) -> List[int]:
        """
        Duyệt postorder bằng đệ quy
        Thứ tự: Trái -> Phải -> Gốc
        """
        result = []
        
        def dfs(node):
            if not node:
                return
            dfs(node.left)            # Duyệt trái
            dfs(node.right)           # Duyệt phải
            result.append(node.val)   # Thăm gốc
        
        dfs(root)
        return result
    
    def postorder_iterative_two_stacks(self, root: Optional[TreeNode]) -> List[int]:
        """
        Duyệt postorder không đệ quy bằng 2 stack
        """
        if not root:
            return []
        
        stack1 = [root]
        stack2 = []
        result = []
        
        # Stack 1: duyệt và đẩy vào stack 2
        while stack1:
            node = stack1.pop()
            stack2.append(node)
            
            if node.left:
                stack1.append(node.left)
            if node.right:
                stack1.append(node.right)
        
        # Stack 2: lấy ra theo thứ tự postorder
        while stack2:
            result.append(stack2.pop().val)
        
        return result
    
    def postorder_iterative_one_stack(self, root: Optional[TreeNode]) -> List[int]:
        """
        Duyệt postorder không đệ quy bằng 1 stack với flag
        """
        if not root:
            return []
        
        result = []
        stack = [(root, False)]  # (node, processed)
        
        while stack:
            node, processed = stack.pop()
            
            if processed:
                result.append(node.val)
            else:
                # Đánh dấu đã xử lý và push lại
                stack.append((node, True))
                # Push con phải trước, trái sau
                if node.right:
                    stack.append((node.right, False))
                if node.left:
                    stack.append((node.left, False))
        
        return result
    
    # ===== TOP-DOWN TRAVERSAL =====
    
    def level_order_traversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        """
        Duyệt theo mức từ trên xuống (Level order)
        """
        if not root:
            return []
        
        result = []
        queue = deque([root])
        
        while queue:
            level_size = len(queue)
            current_level = []
            
            for _ in range(level_size):
                node = queue.popleft()
                current_level.append(node.val)
                
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            
            result.append(current_level)
        
        return result
    
    def top_down_flat(self, root: Optional[TreeNode]) -> List[int]:
        """
        Duyệt top-down trả về 1 list phẳng
        """
        if not root:
            return []
        
        result = []
        queue = deque([root])
        
        while queue:
            node = queue.popleft()
            result.append(node.val)
            
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        
        return result
    
    # ===== BOTTOM-UP TRAVERSAL =====
    
    def bottom_up_traversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        """
        Duyệt theo mức từ dưới lên
        """
        levels = self.level_order_traversal(root)
        return levels[::-1]  # Đảo ngược
    
    def bottom_up_flat(self, root: Optional[TreeNode]) -> List[int]:
        """
        Duyệt bottom-up trả về 1 list phẳng
        """
        levels = self.level_order_traversal(root)
        result = []
        
        for level in reversed(levels):
            result.extend(level)
        
        return result
    
    # ===== CÂY TỔNG QUÁT =====
    
    def preorder_general(self, root: Optional[GeneralTreeNode]) -> List[int]:
        """
        Duyệt preorder cho cây tổng quát
        """
        result = []
        
        def dfs(node):
            if not node:
                return
            result.append(node.val)
            for child in node.children:
                dfs(child)
        
        dfs(root)
        return result
    
    def postorder_general(self, root: Optional[GeneralTreeNode]) -> List[int]:
        """
        Duyệt postorder cho cây tổng quát
        """
        result = []
        
        def dfs(node):
            if not node:
                return
            for child in node.children:
                dfs(child)
            result.append(node.val)
        
        dfs(root)
        return result
    
    def level_order_general(self, root: Optional[GeneralTreeNode]) -> List[List[int]]:
        """
        Duyệt level order cho cây tổng quát
        """
        if not root:
            return []
        
        result = []
        queue = deque([root])
        
        while queue:
            level_size = len(queue)
            current_level = []
            
            for _ in range(level_size):
                node = queue.popleft()
                current_level.append(node.val)
                
                for child in node.children:
                    queue.append(child)
            
            result.append(current_level)
        
        return result
    
    # ===== UTILITY FUNCTIONS =====
    
    def create_sample_tree(self) -> TreeNode:
        """
        Tạo cây mẫu:
                1
               / \
              2   3
             / \   \
            4   5   6
        """
        root = TreeNode(1)
        root.left = TreeNode(2)
        root.right = TreeNode(3)
        root.left.left = TreeNode(4)
        root.left.right = TreeNode(5)
        root.right.right = TreeNode(6)
        return root
    
    def create_sample_general_tree(self) -> GeneralTreeNode:
        """
        Tạo cây tổng quát mẫu:
                1
              / | \
             2  3  4
            /|    |\
           5 6    7 8
        """
        root = GeneralTreeNode(1)
        root.children = [GeneralTreeNode(2), GeneralTreeNode(3), GeneralTreeNode(4)]
        
        root.children[0].children = [GeneralTreeNode(5), GeneralTreeNode(6)]
        root.children[2].children = [GeneralTreeNode(7), GeneralTreeNode(8)]
        
        return root
    
    def get_height(self, root: Optional[TreeNode]) -> int:
        """Tính chiều cao cây"""
        if not root:
            return 0
        return 1 + max(self.get_height(root.left), self.get_height(root.right))
    
    def count_nodes(self, root: Optional[TreeNode]) -> int:
        """Đếm số node trong cây"""
        if not root:
            return 0
        return 1 + self.count_nodes(root.left) + self.count_nodes(root.right)
    
    def is_balanced(self, root: Optional[TreeNode]) -> bool:
        """Kiểm tra cây có cân bằng hay không"""
        def check_balance(node):
            if not node:
                return 0
            
            left_height = check_balance(node.left)
            if left_height == -1:
                return -1
            
            right_height = check_balance(node.right)
            if right_height == -1:
                return -1
            
            if abs(left_height - right_height) > 1:
                return -1
            
            return 1 + max(left_height, right_height)
        
        return check_balance(root) != -1

def print_list(lst, name):
    """In danh sách"""
    print(f"{name}: {lst}")

def print_levels(levels, name):
    """In các mức"""
    print(f"{name}:")
    for i, level in enumerate(levels):
        print(f"  Level {i}: {level}")

# Hàm test chính
def main():
    traverser = TreeTraversal()
    
    print("=== DEMO DUYỆT CÂY NHỊ PHÂN ===")
    
    # Tạo cây mẫu
    root = traverser.create_sample_tree()
    
    print("Cây mẫu:")
    print("        1")
    print("       / \\")
    print("      2   3")
    print("     / \\   \\")
    print("    4   5   6")
    print()
    
    # Test preorder
    pre_rec = traverser.preorder_recursive(root)
    pre_iter = traverser.preorder_iterative(root)
    print_list(pre_rec, "Preorder (Đệ quy)")
    print_list(pre_iter, "Preorder (Không đệ quy)")
    
    # Test postorder
    post_rec = traverser.postorder_recursive(root)
    post_two = traverser.postorder_iterative_two_stacks(root)
    post_one = traverser.postorder_iterative_one_stack(root)
    print_list(post_rec, "Postorder (Đệ quy)")
    print_list(post_two, "Postorder (2 stack)")
    print_list(post_one, "Postorder (1 stack)")
    
    # Test level order
    top_down = traverser.level_order_traversal(root)
    bottom_up = traverser.bottom_up_traversal(root)
    print_levels(top_down, "Top-down traversal")
    print_levels(bottom_up, "Bottom-up traversal")
    
    print()
    
    # Thông tin cây
    print("=== THÔNG TIN CÂY ===")
    print(f"Chiều cao: {traverser.get_height(root)}")
    print(f"Số node: {traverser.count_nodes(root)}")
    print(f"Cân bằng: {'Có' if traverser.is_balanced(root) else 'Không'}")
    
    print()
    
    # Test cây tổng quát
    print("=== DEMO DUYỆT CÂY TỔNG QUÁT ===")
    
    gen_root = traverser.create_sample_general_tree()
    
    print("Cây tổng quát mẫu:")
    print("        1")
    print("      / | \\")
    print("     2  3  4")
    print("    /|    |\\")
    print("   5 6    7 8")
    print()
    
    gen_pre = traverser.preorder_general(gen_root)
    gen_post = traverser.postorder_general(gen_root)
    gen_levels = traverser.level_order_general(gen_root)
    
    print_list(gen_pre, "Preorder cây tổng quát")
    print_list(gen_post, "Postorder cây tổng quát")
    print_levels(gen_levels, "Level order cây tổng quát")

if __name__ == "__main__":
    main()